helix-core/src/main/java/org/apache/helix/common/caches/ParticipantStateCache.java

```diff
@@ -0,0 +1,203 @@
+        List<String> currentStateNames =
+            accessor.getChildNames(keyBuilder.currentStates(instanceName, sessionId));
+
+        for (String currentStateName : currentStateNames) {
+          participantStateKeys
+              .add(keyBuilder.currentState(instanceName, sessionId, currentStateName));
+        }
+      } else {
+        List<String> resourceNames = new ArrayList<>();
+        for (String customizedStateType : restrictedKeys) {
+          resourceNames = accessor
+              .getChildNames(keyBuilder.customizedStates(instanceName, customizedStateType));
+          for (String resourceName : resourceNames) {
+            participantStateKeys
+                .add(keyBuilder.customizedState(instanceName, customizedStateType, resourceName));
+          }
+        }
+      }
+    }
+    // All new entries from zk not cached locally yet should be read from ZK.
+    Set<PropertyKey> reloadKeys = new HashSet<>(participantStateKeys);
+    reloadKeys.removeAll(_participantStateCache.keySet());
+
+    Set<PropertyKey> cachedKeys = new HashSet<>(_participantStateCache.keySet());
+    cachedKeys.retainAll(participantStateKeys);
+
+    Set<PropertyKey> reloadedKeys = new HashSet<>();
+    Map<PropertyKey, T> newStateCache = Collections.unmodifiableMap(refreshProperties(accessor,
+        reloadKeys, new ArrayList<>(cachedKeys), _participantStateCache, reloadedKeys));
+
+    if (snapshotEnabled) {
+      refreshSnapshot(newStateCache, _participantStateCache, reloadedKeys);
+    }
+
+    _participantStateCache = newStateCache;
+
+    if (LOG.isDebugEnabled()) {
+      LogUtil.logDebug(LOG, genEventInfo(), "# of participant state reload: " + reloadKeys.size()
+          + ", skipped:" + (participantStateKeys.size() - reloadKeys.size()) + ". took "
+          + (System.currentTimeMillis() - start) + " ms to reload new participant states for cluster: "
+          + _controlContextProvider.getClusterName() + "and state: " + this.getClass().getName());
+    }
+  }
+
+  protected void refreshSnapshot(Map<PropertyKey, T> newStateCache,
+      Map<PropertyKey, T> participantStateCache, Set<PropertyKey> reloadedKeys) {
+  }
+
+  /**
+   * Return the whole participant state map.
+   * @return
+   */
+  public Map<String, Map<String, Map<String, T>>> getParticipantStatesMap() {
+    return Collections.unmodifiableMap(_participantStateMap);
+  }
+
+  /**
+   * Return all participant states for a certain level1 key.
+   * @param keyLevel1
+   * @return
+   */
+  public Map<String, Map<String, T>> getParticipantStates(String keyLevel1) {
+    if (!_participantStateMap.containsKey(keyLevel1)) {
+      return Collections.emptyMap();
+    }
+    return Collections.unmodifiableMap(_participantStateMap.get(keyLevel1));
+  }
+
+  /**
+   * Provides the participant state map for a certain level1 key and level2 key
+   * @param keyLevel1
+   * @param keyLevel2
+   * @return
+   */
+  public Map<String, T> getParticipantState(String keyLevel1,
```

Comment:
Optional, both getParticipantStates() methods could be protected and let the children class wrap them and implement with more meaningful public methods.

Reply:
I changed the key names here. I feel it's ok to leave it here as it only returns some mapping.